File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = statement match {
    case statement if isShouting(statement) => "Whoa, chill out!"
    case statement if isQuestion(statement) => "Sure."
    case statement if isSilent(statement) => "Fine. Be that way!"
    case _ => "Whatever."
  }

  private def isShouting(statement: String) =
    isUppperCase(statement) && isNotNum(statement)

  private def isUppperCase(statement: String) =
    statement.toUpperCase() == statement

  private def isNotNum(statement: String) =
    statement.exists(_.isLetter)

  private def isQuestion(statement: String) =
    removeWhiteSpace(statement).endsWith("?")

  private def isSilent(statement: String) =
    removeWhiteSpace(statement).isEmpty()
  
  private def removeWhiteSpace(statement: String) =
    statement.trim
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      question_in_capitals?(input) ->
        "Calm down, I know what I'm doing!"

      question?(input) ->
        "Sure."

      silence?(input) ->
        "Fine. Be that way!"

      shouting?(input) ->
        "Whoa, chill out!"

      true ->
        "Whatever."
    end
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  defp shouting?(input) do
    input == String.upcase(input) && String.match?(input, ~r/[^\d, ]+/)
  end

  defp silence?(input) do
    String.trim(input) == ""
  end

  defp question_in_capitals?(input) do
    question?(input) && input == String.upcase(input) && input =~ ~r/[[:alpha:]]/
  end
end
